src/app/reporting/ReportScheduler.h

```diff
@@ -0,0 +1,150 @@
+        /// handler state, and minimal time interval last since last report has elapsed, or the maximal time interval since last
+        /// report has elapsed
+        bool IsReportableNow() const
+        {
+            // TODO: Add flags to allow for test to simulate waiting for the min interval or max intrval to elapse when integrating
+            // the scheduler in the ReadHandler
+            Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            return (mReadHandler->IsGeneratingReports() &&
+                    ((now >= mMinTimestamp && mReadHandler->IsDirty()) || now >= mMaxTimestamp));
+        }
+
+        void SetIntervalTimeStamps(ReadHandler * aReadHandler)
+        {
+            uint16_t minInterval, maxInterval;
+            aReadHandler->GetReportingIntervals(minInterval, maxInterval);
+            Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            mMinTimestamp = now + System::Clock::Seconds16(minInterval);
+            mMaxTimestamp = now + System::Clock::Seconds16(maxInterval);
+        }
+
+        void RunCallback() { mCallback(); }
+
+        Timestamp GetMinTimestamp() const { return mMinTimestamp; }
+        Timestamp GetMaxTimestamp() const { return mMaxTimestamp; }
+
+    private:
+        TimerDelegate * mTimerDelegate;
+        TimerCompleteCallback mCallback;
+        ReadHandler * mReadHandler;
+        Timestamp mMinTimestamp;
+        Timestamp mMaxTimestamp;
+    };
+
+    ReportScheduler(TimerDelegate * aTimerDelegate) : mTimerDelegate(aTimerDelegate) {}
+    /**
+     *  Interface to act on changes in the ReadHandler reportability
+     */
+    virtual ~ReportScheduler() = default;
+
+    /// @brief Check if a ReadHandler is scheduled for reporting
+    virtual bool IsReportScheduled(ReadHandler * aReadHandler) = 0;
+    /// @brief Check whether a ReadHandler is reportable right now, taking into account its minimum and maximum intervals.
+    /// @param aReadHandler read handler to check
+    bool IsReportableNow(ReadHandler * aReadHandler) { return FindReadHandlerNode(aReadHandler)->IsReportableNow(); };
+    /// @brief Check if a ReadHandler is reportable without considering the timing
+    bool IsReadHandlerReportable(ReadHandler * aReadHandler) const
+    {
+        return aReadHandler->IsGeneratingReports() && aReadHandler->IsDirty();
+    }
+
+    /// @brief Get the number of ReadHandlers registered in the scheduler's node pool
+    size_t GetNumReadHandlers() const { return mNodesPool.Allocated(); }
+
+protected:
+    friend class chip::app::reporting::TestReportScheduler;
+
+    /// @brief Find the ReadHandlerNode for a given ReadHandler pointer
+    /// @param [in] aReadHandler ReadHandler pointer to look for in the ReadHandler nodes list
+    /// @return Node Address if node was found, nullptr otherwise
+    ReadHandlerNode * FindReadHandlerNode(const ReadHandler * aReadHandler)
```

Comment:
As a followup: We should consider moving some of these now-inlined things into an actual .cpp file....

Reply:
issue here: https://github.com/project-chip/connectedhomeip/issues/27927